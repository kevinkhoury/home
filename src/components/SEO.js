import React, { useContext } from 'react'
import moment from 'moment'
import { Helmet } from 'react-helmet'
import PropTypes from 'prop-types'
import { featuredImagePropTypes } from '../proptypes'
import { isMobile } from 'react-device-detect'

import { useStructuredData, useSiteData } from '../hooks'
import { ThemeOptionsContext } from '../context/ThemeOptions'
import { addTrailingSlash } from '../utils'

const buildImage = (img, siteUrl) => {
  const url = `${siteUrl}/img/${img.childImageSharp.fluid.originalName}`
  const { width, height } = img.childImageSharp.original
  return { url, width, height }
}

const Moment = moment().constructor

const SEO = ({
  pageTitle,
  metaDescription,
  featuredImage,
  slug,
  date,
  dateModified,
  templateKey,
  schemaType,
}) => {
  const {
    name,
    siteName,
    jobTitle,
    siteUrl,
    socialLinks: { twitter, facebook, linkedin, pinterest, instagram },
    fallbackImage,
  } = useSiteData()

  const image =
    featuredImage && featuredImage.src ? featuredImage.src : fallbackImage

  const schemaData = {
    templateKey,
    headline: pageTitle,
    description: metaDescription,
    url: `${siteUrl}${addTrailingSlash(slug)}`,
    fallbackImage: buildImage(fallbackImage, siteUrl),
    pageType:
      templateKey && templateKey.indexOf('blog-post') !== -1
        ? 'article'
        : 'webpage',
    datePublished: date.format('YYYY-MM-DD'),
    dateModified: dateModified.format('YYYY-MM-DD'),
    name,
    jobTitle,
    siteName,
    schemaType,
    sameAs: [twitter, facebook, linkedin, pinterest, instagram].filter(
      (item) => !!item && !!item.url && item.url.length,
    ),
  }

  schemaData.image = !!image
    ? buildImage(image, siteUrl)
    : schemaData.fallbackImage

  const twitterHandle =
    twitter && twitter.url ? '@' + twitter.url.split('twitter.com/')[1] : null

  const ldjson = useStructuredData(schemaData)

  const { colorScheme, fontScheme } = useContext(ThemeOptionsContext)

  const htmlClasses = [
    colorScheme,
    fontScheme,
    typeof window !== 'undefined' && isMobile ? 'no-mouse' : '',
  ]
    .filter((x) => x)
    .join(' ')

  return (
    <Helmet>
      <html lang="en" className={htmlClasses} />
      <title>{pageTitle}</title>
      <meta
        name="robots"
        content="max-snippet:-1, max-image-preview:large, max-video-preview:-1"
      />
      <meta name="description" content={metaDescription} />
      <meta property="og:type" content={schemaData.pageType} />
      <meta property="og:title" content={pageTitle} />
      <meta property="og:description" content={metaDescription} />
      <meta property="og:url" content={schemaData.url} />
      <meta property="og:image" content={schemaData.image.url} />
      {schemaData.pageType === 'article' && (
        <meta property="article:author" content={name} />
      )}
      <meta name="twitter:card" content="summary_large_image" />
      {!!twitterHandle && <meta name="twitter:site" content={twitterHandle} />}
      <meta name="twitter:title" content={pageTitle} />
      <meta name="twitter:description" content={metaDescription} />
      <meta name="twitter:image" content={schemaData.image.url} />
      <script type="application/ld+json">{ldjson}</script>
    </Helmet>
  )
}

export const seoPropTypes = {
  pageTitle: PropTypes.string.isRequired,
  metaDescription: PropTypes.string.isRequired,
  featuredImage: featuredImagePropTypes,
  slug: PropTypes.string.isRequired,
  date: PropTypes.instanceOf(Moment).isRequired,
  dateModified: PropTypes.instanceOf(Moment).isRequired,
  templateKey: PropTypes.string.isRequired,
  schemaType: PropTypes.string.isRequired,
}

SEO.propTypes = seoPropTypes

export default SEO
