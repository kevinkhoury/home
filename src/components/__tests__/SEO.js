import React, { useContext as useContextMock } from 'react'
import moment from 'moment'
import renderer from 'react-test-renderer'
import { render, screen, waitFor } from '@testing-library/react'
import SEO from '../SEO'
import {
  useSiteData as useSiteDataMock,
  useStructuredData as useStructuredDataMock,
  useStructuredData,
} from '../../hooks'

const DEFAULT_PROPS = {
  pageTitle: 'pageTitle',
  metaDescription: 'metaDescription',
  featuredImage: {
    src: {
      childImageSharp: {
        fluid: {
          originalName: 'foobar.jpg',
        },
        original: {
          height: 100,
          width: 200,
        },
      },
    },
    d: {
      childImageSharp: {},
    },
    m: {
      childImageSharp: {},
    },
    alt: 'imgalt',
    caption: 'imgcaption',
  },
  slug: '/foobar',
  date: moment('20200101', 'YYYYMMDD'),
  dateModified: moment('20200101', 'YYYYMMDD'),
  templateKey: 'templatekey',
  schemaType: 'schematype',
}

jest.mock('../../context/ThemeOptions', () => () => 'ThemeOptionsContext')
jest.mock('react-device-detect', () => ({ isMobile: jest.fn() }))
jest.mock('../../hooks', () => ({
  useSiteData: jest.fn(),
  useStructuredData: jest.fn(),
}))
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}))

function mockSocialLink(show) {
  return {
    url: 'https://foobar.com',
    show: !!show,
  }
}

beforeEach(() => {
  useContextMock.mockImplementation(() => ({
    colorScheme: 'colorScheme',
    fontScheme: 'fontScheme',
  }))
  useSiteDataMock.mockImplementation(() => ({
    name: 'name',
    siteName: 'siteName',
    jobTitle: 'jobTitle',
    siteUrl: 'siteUrl',
    socialLinks: {
      twitter: { url: 'https://twitter.com/foobar', show: true },
      facebook: mockSocialLink(true),
      linkedin: mockSocialLink(false),
      pinterest: mockSocialLink(true),
      instagram: mockSocialLink(false),
    },
    fallbackImage: {
      childImageSharp: {
        original: {
          height: 100,
          width: 200,
        },
        fluid: {
          originalName: 'fallback.jpg',
        },
      },
    },
  }))
  useStructuredData.mockImplementation((props) => JSON.stringify(props))
})

afterEach(() => {
  jest.clearAllMocks()
})

describe('SEO', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<SEO {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('creates a document head with all parameters', async () => {
    render(<SEO {...DEFAULT_PROPS} />)
    await waitFor(() => expect(document.title).toEqual('pageTitle'))
    expect(
      document.head
        .querySelector('[name="description"]')
        .getAttribute('content'),
    ).toEqual('metaDescription')
    expect(
      document.head
        .querySelector('[property="og:type"]')
        .getAttribute('content'),
    ).toEqual('webpage')
    expect(
      document.head
        .querySelector('[property="og:title"]')
        .getAttribute('content'),
    ).toEqual('pageTitle')
    expect(
      document.head
        .querySelector('[property="og:description"]')
        .getAttribute('content'),
    ).toEqual('metaDescription')
    expect(
      document.head
        .querySelector('[property="og:url"]')
        .getAttribute('content'),
    ).toEqual('siteUrl/foobar/')
    expect(
      document.head.querySelector('[property="article:author"]'),
    ).toBeNull()
    expect(
      document.head
        .querySelector('[property="og:image"]')
        .getAttribute('content'),
    ).toEqual('siteUrl/img/foobar.jpg')
    expect(
      document.head
        .querySelector('[name="twitter:site"]')
        .getAttribute('content'),
    ).toEqual('@foobar')
    expect(
      document.head
        .querySelector('[name="twitter:title"]')
        .getAttribute('content'),
    ).toEqual('pageTitle')
    expect(
      document.head
        .querySelector('[name="twitter:description"]')
        .getAttribute('content'),
    ).toEqual('metaDescription')
    expect(
      document.head
        .querySelector('[name="twitter:image"]')
        .getAttribute('content'),
    ).toEqual('siteUrl/img/foobar.jpg')
    expect(
      document.head.querySelector('[type="application/ld+json"]').innerHTML,
    ).toMatch(/"sameAs":\[/)
  })
})

describe('buildImage', () => {
  it('formats image data for schema into `url`, `width`, and `height`', () => {
    const buildImage = SEO.__get__('buildImage')
    const { siteUrl, fallbackImage } = useSiteDataMock()
    const result = buildImage(fallbackImage, siteUrl)
    expect(result.url).toBe('siteUrl/img/fallback.jpg')
    expect(result.width).toBe(200)
    expect(result.height).toBe(100)
  })
})
