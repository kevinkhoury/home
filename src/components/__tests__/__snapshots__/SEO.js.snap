// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO snapshot renders properly 1`] = `null`;
